File: src/neetcode/Java/_39CombinationSum.java
package neetcode.Java;

import java.util.ArrayList;
import java.util.List;

public class _39CombinationSum {
    List<List<Integer>> list = new ArrayList<>();

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        ArrayList<Integer> l = new ArrayList<>();
        Helper(0, candidates, target, l);
        return list;
    }

    public void Helper(int n, int[] num, int tar, ArrayList<Integer> l) {
        if (tar == 0) {
            list.add(new ArrayList<>(l));
            return;
        }
        for (int i = n; i < num.length; i++) {
            if (tar - num[i] >= 0) {
                l.add(num[i]);
                Helper(i, num, tar - num[i], l);
                l.remove(l.size() - 1);
            }
        }
    }

    public static void main(String[] args) {
        _39CombinationSum obj = new _39CombinationSum();
        int[] num = {2, 3, 6, 7};
        int target = 7;
        System.out.println(obj.combinationSum(num, target));
    }
}

File: src/neetcode/kotlin/_39CombinationSum.kt
package neetcode.kotlin

class _39CombinationSum {
    var list: MutableList<List<Int>> = ArrayList()

    fun combinationSum(candidates: IntArray, target: Int): List<List<Int>> {
        val l = ArrayList<Int>()
        Helper(0, candidates, target, l)
        return list
    }

    fun Helper(n: Int, num: IntArray, tar: Int, l: ArrayList<Int>) {
        if (tar == 0) {
            list.add(ArrayList(l))
            return
        }
        for (i in n until num.size) {
            if (tar - num[i] >= 0) {
                l.add(num[i])
                Helper(i, num, tar - num[i], l)
                l.removeAt(l.size - 1)
            }
        }
    }

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            val obj = _39CombinationSum()
            val num = intArrayOf(2, 3, 6, 7)
            val target = 7
            println(obj.combinationSum(num, target))
        }
    }
}
